### PR TITLE
Fix recurrent op not update grade issue

### PR DESCRIPTION
**Pull request by zhaoyuchen2018**
Remove same name parameter in recurrent op.
test=develop
https://github.com/PaddlePaddle/Paddle/issues/18803

### CODE REVIEW

File: paddle/fluid/operators/recurrent_op.cc [hunk 1]
```diff
@@ -469,16 +469,36 @@ class RecurrentGradOp : public RecurrentBase {
...
+          // add inside grad variable to outside grad variable
+          // if subblock's variable affect parent block varieble.
+          if (inside_grad_name.compare(pg_names[param_id]) != 0) {
+            const framework::Tensor &cur_var =
+                cur_scope.FindLocalVar(pg_names[param_id])
+                    ->Get<framework::LoDTensor>();
+            auto &pre_var = const_cast<framework::LoDTensor &>(
+                scope.FindLocalVar(inside_grad_name)
+                    ->Get<framework::LoDTensor>());
+            pre_var.ShareDataWith(cur_var);
```
Comment (phlrain) [inline paddle/fluid/operators/recurrent_op.cc]: 我们不是不建议在op 内部使用share data with 操作

Comment (zhaoyuchen2018) [reply paddle/fluid/operators/recurrent_op.cc]: 有什么副作用么，如果不用的话那就需要copy数据

Comment (phlrain) [reply paddle/fluid/operators/recurrent_op.cc]: 我去了解下，我对share data with理解有误，这块我没有问题了。

File: paddle/fluid/operators/recurrent_op.cc [hunk 2]
```diff
@@ -377,16 +377,35 @@ void RecurrentGradOp::RunImpl(const framework::Scope &scope,
...
-        auto sum_op = framework::OpRegistry::CreateOp(
-            "sum", {{"X", {pg_names[param_id], new_inside_name}}},
-            {{"Out", {pg_names[param_id]}}},
-            framework::AttributeMap{{"use_mkldnn", {false}}});
-        sum_op->Run(cur_scope, place);
-
-        cur_scope.Rename(new_inside_name, inside_grad_name);
+        // add inside grad variable to outside grad variable
+        // if subblock's variable affect parent block varieble.
+        if (inside_grad_name.compare(pg_names[param_id]) != 0) {
```
Comment (Xreki) [inline paddle/fluid/operators/recurrent_op.cc]: inside_grad_name == pg_names[param_id]

File: paddle/fluid/operators/recurrent_op.cc [hunk 3]
```diff
@@ -377,16 +377,35 @@ void RecurrentGradOp::RunImpl(const framework::Scope &scope,
...
+
+          // sum gradient
+          auto sum_op = framework::OpRegistry::CreateOp(
+              "sum", {{"X", {pg_names[param_id], new_inside_name}}},
+              {{"Out", {pg_names[param_id]}}},
+              framework::AttributeMap{{"use_mkldnn", {false}}});
+          sum_op->Run(cur_scope, place);
+
+          cur_scope.Rename(new_inside_name, inside_grad_name);
+        }
```
Comment (Xreki) [inline paddle/fluid/operators/recurrent_op.cc]: 感觉这里的逻辑有些奇怪，请@chengduoZH 来看下吧

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py [hunk 4]
```diff
@@ -456,5 +456,145 @@ def create_rnn_op(self):
...
+            rnn_out = self.create_rnn_op()
+            self.output = layers.mean(rnn_out)
+
+    def create_rnn_op(self):
+        x = layers.data(
+            shape=[self.sent_len, self.batch_size, self.input_dim],
+            dtype='float32',
+            name='x',
+            append_batch_size=False)
+        x.stop_gradient = False
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_recurrent_op.py:547]: Why do you add `x.stop_gradient = False` here?

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py [hunk 5]
```diff
@@ -456,5 +456,145 @@ def create_rnn_op(self):
...
+        x += 0.0
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_recurrent_op.py]: Why?

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py [hunk 6]
```diff
@@ -456,5 +456,145 @@ def create_rnn_op(self):
...
+
+        emb = layers.data(
+            name='emb',
+            shape=[self.sent_len, self.batch_size, self.input_dim],
+            dtype='float32',
+            append_batch_size=False)
+        emb.stop_gradient = False
+        emb += 0.0
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_recurrent_op.py]: Why???

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py [hunk 7]
```diff
@@ -456,5 +456,145 @@ def create_rnn_op(self):
...
+            dtype='float32',
+            name='w1',
+            append_batch_size=False)
+        w1.stop_gradient = False
+        w2 = layers.data(
+            shape=[self.input_dim * 2, self.input_dim],
+            dtype='float32',
+            name='w2',
+            append_batch_size=False)
+        w2.stop_gradient = False
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_recurrent_op.py:567]: It seems that not need to set stop_gradient here.

File: paddle/fluid/operators/recurrent_op.cc (hunk 2 above)
Comment (chengduoZH) [inline paddle/fluid/operators/recurrent_op.cc]: Why?

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py (hunk 4 above)
Comment (zhaoyuchen2018) [reply python/paddle/fluid/tests/unittests/test_recurrent_op.py:547]: https://github.com/PaddlePaddle/Paddle/blob/8cb4d6535c5e7d6eac3b17330c1633d539e6a6a2/python/paddle/fluid/tests/unittests/test_recurrent_op.py#L547

Copy from original file

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py (hunk 5 above)
Comment (zhaoyuchen2018) [reply python/paddle/fluid/tests/unittests/test_recurrent_op.py]: Come from Hongyu's test case, seems another problem

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py (hunk 7 above)
Comment (zhaoyuchen2018) [reply python/paddle/fluid/tests/unittests/test_recurrent_op.py:567]: why? w2 should also compute gradient

File: paddle/fluid/operators/recurrent_op.cc (hunk 2 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/recurrent_op.cc]: find the variables have same name in both block so they can share data

Comment (zhaoyuchen2018) [reply paddle/fluid/operators/recurrent_op.cc]: 这个两个结果是一样的

File: paddle/fluid/operators/recurrent_op.cc (hunk 3 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/recurrent_op.cc]: 这里是原始的逻辑。我理解是因为前面sum使用了new_inside_name，所以后面就用new_inside_name来替换原始的name了。

**Review by phlrain — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/recurrent_op.cc [hunk 8]
```diff
@@ -377,16 +377,35 @@ void RecurrentGradOp::RunImpl(const framework::Scope &scope,
...
-
-        cur_scope.Rename(new_inside_name, inside_grad_name);
+        // add inside grad variable to outside grad variable
+        // if subblock's variable affect parent block varieble.
+        if (inside_grad_name.compare(pg_names[param_id]) != 0) {
+          const framework::Tensor &cur_var =
+              cur_scope.FindLocalVar(pg_names[param_id])
+                  ->Get<framework::LoDTensor>();
+          auto &pre_var = const_cast<framework::LoDTensor &>(
+              scope.FindVar(inside_grad_name)->Get<framework::LoDTensor>());
```
Comment (Xreki) [inline paddle/fluid/operators/recurrent_op.cc]: Can you call `GetMutable` to avoid `const_cast`?

**Comment by chengduoZH** [comment]
@zhaoyuchen2018  Just need to fix this line:
https://github.com/PaddlePaddle/Paddle/blob/bce72c7feabbb9b82d736631c1e38b5709e27910/python/paddle/fluid/layers/control_flow.py#L565

->

```
parameters = [parent_block.var(name) for name in set(params)] 
```

**Review by chengduoZH — CHANGES_REQUESTED** [review]
This is an error fix.

**Review by chengduoZH — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/test_recurrent_op.py [hunk 9]
```diff
@@ -456,5 +456,143 @@ def create_rnn_op(self):
...
+            concat_in = layers.concat([step_in, pre_h], 1)
+            new_h = layers.matmul(concat_in, w2)
+            new_h = layers.unsqueeze(new_h, [1])
+            new_h, _ = dot_attention(new_h, y)
+            new_h = layers.squeeze(new_h, [1])
+
+            rnn.update_memory(pre_h, new_h)
+            rnn.step_output(new_h)
+
+        return rnn()
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_recurrent_op.py:594]: This unit test should be simplified.

**Pull request merged**